Python/ceval.c

```diff
@@ -4651,6 +4703,11 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
                 _PyFrame_SetStackPointer(frame, stack_pointer);
                 new_frame->previous = frame;
                 cframe.current_frame = frame = new_frame;
+                if (call_shape.init_pass_self) {
```

Comment:
What happens if `init_pass_self` is set, but `function` is not a normal Python function, e.g. a generator function?

Reply:
I'm unsure how that would work? Isn't `__init__` being a generator invalid?

Reply:
It's invalid but it shouldn't crash the interpreter.

Reply:
Interesting, I didn't expect this result at all!

I expected a refleak, but it doesn't and instead treats the generator as the `self` object.
```python
from sys import gettotalrefcount as gtr

class Y:
 def __init__(s):
  yield 1

for _ in range(20):
 gtr()
 try:
  Y()
 except TypeError as e:
  print(e)


__init__() should return None, not 'generator'
39767
__init__() should return None, not 'generator'
39767
__init__() should return None, not 'generator'
39767
__init__() should return None, not 'generator'
39767
__init__() should return None, not 'generator'
39767
__init__() should return None, not 'generator'
39767
__init__() should return None, not 'generator'
39767
<generator object Y.__init__ at 0x000001FB271A4E50>
39775
<generator object Y.__init__ at 0x000001FB271A4D70>
39775
<generator object Y.__init__ at 0x000001FB271A4D70>
39775
<generator object Y.__init__ at 0x000001FB271A4D70>
39775
<generator object Y.__init__ at 0x000001FB271A4D70>
39775
<generator object Y.__init__ at 0x000001FB271A4D70>
39775
<generator object Y.__init__ at 0x000001FB271A4D70>
39775
```
The fix should be to check for generators and drop them.

Reply:
Fixed. Thank you both!